Nest 2nd version - 02 nest (#5)
raven-os/getting-started: nest/03.nest.md

# 2 nest
In this chapter, we will detail how to use the binary `nest`.

> As explained before, this binary is mainly targeted towards newcomers.

Reviewer: ```suggestion
As explained before, this binary is mainly designed for ease of use.
```
This is a wild suggestion, feel free to ignore, but I could imagine that many people may stick to this binary if they don't need more features. The next sentence reinforces my feeling.

However, anyone can use it.

## 2.1 nest options
`nest` has three options available:

* `-h` or `--help`, which prints the help information of `nest`, consisting of a short explanation for each of `nest`'s options and commands.
* `-v`, which sets the level of verbosity for `nest`. Basically, how much `nest` will explain what it's doing in the terminal.
* `-V` or `--version`, which prints the version of `nest` you are currently using.

To use them, simply type `nest` followed by any option. For example, `nest -h`.

## 2.2 nest commands

> ### 2.2.1 Updating the list of available packages

Reviewer: ```suggestion
### 2.2.1 Updating the list of available packages with nest pull
```
I'd put the command in the title

Author: I'm guessing this also applies to the other titles

> If you have used the option `-h` or `--help` already, you know that this command is `nest pull`.

Reviewer: Sounds condescending. I'd remove this.

This command synchronizes the cache of local packages with the one on Raven's server.
In other words, `nest pull` updates the list of available packages for you to install with Nest.
You have to run this command frequently (at least before installing or upgrading a package) in order to have an up to date cache.
Otherwise, you might not be able to install new packages, or new versions of packages that were uploaded after the last time you ran `nest pull`.

#### Common issues
* `error: unable to pull repository stable: no working mirror found`

[//]: # (TODO: add link to the section 3.1.3)
Basically, this error means that `nest` couldn't pull the repository `stable`.
The most common reason is that there's an issue with your Internet connection.
It might also be that all the mirrors in the repositories configuration file are down, so you need to add working mirrors to the list.
You can have more information about this file and mirrors in [the section 3.1.3]().

### 2.2.2 Installing and uninstalling a package
#### Installing a package
[//]: # (TODO: add link to the section 1.4)
`nest install` is the command to install a package.
To do so, run `nest install <package>` in a terminal.

> Don't forget to use the syntax explained in [the introduction]() for the package name!

Reviewer: Maybe refer to the exact section?

`nest` will print an error message if more than one package matches the syntax you wrote.

> In that case, you can go on [the official Raven-OS website](https://stable.raven-os.org) and change the syntax accordingly.

Reviewer: There seems to be a missing previous step, where you explain that you need to refine your package name. If the way to do so isn't obvious, *then* go to the website.

For example, the output of `nest install stable::sys-bin/ncurses6-bin` is as followed:
```
$ nest install ncurses6-bin
4 pending transactions:

 install stable::sys-lib/libncurses6-term#6.1.0
 install stable::sys-lib/libc6#2.28.0
 install stable::sys-lib/libncursestw6#6.1.0
 install stable::sys-bin/ncurses6-bin#6.1.0

Would you like to apply these transactions? [Yes/no]
```

> As you can see, the last package listed above is the package you initially wanted to install, and above it are all the dependencies that it has and that are not already installed on the OS.

Reviewer: ```suggestion
As you can see, the last package listed above is the package you initially asked to install, and above it are all the dependencies that it has and that aren't already installed.
```

You will then be prompted with the question `Would you like to apply these transactions?`, which you can answer with either '*yes*', '*Yes*', '*Y*', '*y*', or '*no*', '*No*', '*n*', '*N*'.

An important thing to note, is that when Nest (and most CLI programs) ask you a closed question like this one, the default answer usually begins with an uppercase letter.

> So in this case, just pressing enter is equivalent to typing `yes` and then pressing enter.

Reviewer: I hope this isn't really the case in the binary. A "yes" default is a bad idea in this case.

Reviewer: @Arignir

Reviewer 2: It is the case. That's debatable, I don't think it's a huge deal but I understand your concerns. We'll discuss it in the next team meeting.

#### Uninstalling a package
The command to uninstall a package is `nest uninstall <package>`.

It will remove the given package, its binaries, libraries, manual pages...
Basically everything that was installed with it, but it will keep all the files generated by the package itself.
So, if this package was a text editor for example, the files you created or edited with it won't be altered if you uninstall the package.

> However, this won't remove the package's dependencies, because they might still be used by another installed package.

Reviewer: ```suggestion
However, this may not remove all the package's dependencies, because some of them might still be used by another installed package.
```
(At least I hope dependencies don't stay forever)

#### Common issues
* `error: no package found for requirement <package>`

This error could mean that you want to install a package that isn't in Nest's cache, so, `nest` can't find it and install it.
Try running `nest pull` and then running your command again.

Another possible explanation is that the package you wanted to install can't be found.
You should check the package name syntax, in case you mistyped the repository, the category, the package name itself or the version.

> You can always check [on Raven-OS website](https://stable.raven-os.org/) as mentioned above, to have not make any mistakes.

Reviewer: ```suggestion
You can always check [on Raven-OS website](https://stable.raven-os.org/) as mentioned above, to make sure you haven't made any mistakes.
```

* `error: unable to find an installed package matching <package>`

The package you wanted to uninstall can't be found.

> You have to check the package name syntax for any typo, and that you already have this package installed.

Reviewer: Maybe link to a way to list installed packages ? Idk if nest supports this.

Author: it does not, unless it's a secret I'm not aware of yet (@Arignir )

Reviewer 2: It has been added at the end of last rush. The command is `nest list`, with an optional parameter `--with-deps`.

Btw, you should ping @doom for questions that are related to nest-cli directly, I'm only the maintainer of the algorithm behind the resolution of dependencies (and even that, I whisper the algorithm to @doom and he writes it).

> ### 2.3 Summary

Reviewer: I'd expect this section to recall everything that was explained.
Something like:
"We saw how to install and uninstall dependencies using nest install and nest uninstall respectively [...]"
This allows the reader to make sure he got everything, jog his short-term memory and increase the chances of someone reading diagonally to get the important info he's looking for.
Then you can lead to the next chapter using something like:
"Now, that you are familiar with nest, the next chapter will explain some advanced concepts [...]"

> This concludes the part of this getting-started dedicated to `nest`!

Reviewer: This sentence sounds very weird. I can't find an alternative right now but I'll check back later.

Author: what about
```suggestion
This concludes the chapter dedicated to `nest`!
```

Reviewer: Sounds far better.

> To be followed are some advanced concepts that are useful, but not necessary, to use `nest`.

Reviewer: ```suggestion
The next chapter explains some advanced concepts that are useful, but not necessary, to use `nest`.
```

> The readers interested in using `finest` should read this next chapter as well, because those concepts are important in order to use `finest`.

Reviewer: ```suggestion
Readers interested in using `finest` should read the next chapter as well, because those concepts are important in order to use `finest`.
```

> # 2 nest

Reviewer 2: All the titles state this is chapter 2, while the filename suggests it's chapter 3.

> In this chapter, we will detail how to use the binary `nest`.

Reviewer 2: ```suggestion
In this chapter, we will detail how to use the command `nest`.
```

> As explained before, this binary is mainly designed for ease of use.

Reviewer 2: ```suggestion
As explained before, this command is mainly designed for ease of use.
```

However, anyone can use it.

> ## 2.1 nest options

Reviewer 2: I'm not sure this section is very useful. There's not so much interest in its content, and it assumes (by it's last sentence about how to form a valid command) that the user haven't typed any command before.

While it's great to start slowly, maybe that's a bit too slow :p We can assume the user has a tiny bit of user experiance with command line and terminals before.

`nest` has three options available:

* `-h` or `--help`, which prints the help information of `nest`, consisting of a short explanation for each of `nest`'s options and commands.

> * `-v`, which sets the level of verbosity for `nest`. Basically, how much `nest` will explain what it's doing in the terminal.

Reviewer 3: This sentence implies that there are several levels of verbosity that can be set, but doesn't really explain how to choose one over another

Reviewer 2: (Would be cool to actually use the flag, now that it's there :o)

* `-V` or `--version`, which prints the version of `nest` you are currently using.

To use them, simply type `nest` followed by any option. For example, `nest -h`.

## 2.2 nest commands
### 2.2.1 Updating the list of available packages with `nest pull`

> This command synchronizes the cache of local packages with the one on Raven's server.

Reviewer 2: Instead of starting by saying what the command does and then why it is useful, maybe explain first the problematic and then the solution.

Something like "Nest needs the list of available packages to know what to do when an install is requested ... you need to do that often for that and that reason ... if you don't horrible things might happen like ... `nest pull` is used for that purpose ..."

What do you think?

Reviewer: I'd let the first sentence define what it is and afterwards add a rational. This allows user that already know the "problem" to skip it (mentally).

Author: that makes sense, I guess it should be that way for the other commands too, when possible
I'm not sure to understand what you meant by 'rational' @Gaspard--

Reviewer: >Something like "Nest needs the list of available packages to know what to do when an install is requested ... you need to do that often for that and that reason ... if you don't horrible things might happen like ... nest pull is used for that purpose ..."

This is a rational, it explains why the command is needed.

Reviewer 2: Okay, why not =)

In other words, `nest pull` updates the list of available packages for you to install with Nest.
You have to run this command frequently (at least before installing or upgrading a package) in order to have an up to date cache.
Otherwise, you might not be able to install new packages, or new versions of packages that were uploaded after the last time you ran `nest pull`.

> 

Reviewer 2: Please include a complete example of the command, with the associated output.

> #### Common issues

Reviewer 2: Instead of having a sperate `Common issues` section, maybe use the infobox we discussed (Unless the `Common issues` section is long, but that's not the case here)?

What do you think of that?

Author: good idea, I totally forgot about it this since the last meeting we've had D:

Author: I was thinking of using an infobox just before the Common Issues section to say how important it is that the local cache is kept up-to-date. But by doing that, there would be two infoboxes in a row (that one, and the common issues one), which might look a bit weird. What do you think about this?

Reviewer 2: I do not think it's weird to have two infobox in a row, especially if they are of different kind (and therefore will have different colors)

* `error: unable to pull repository stable: no working mirror found`

[//]: # (TODO: add link to the section 3.1.3)
Basically, this error means that `nest` couldn't pull the repository `stable`.
The most common reason is that there's an issue with your Internet connection.
It might also be that all the mirrors in the repositories configuration file are down, so you need to add working mirrors to the list.
You can have more information about this file and mirrors in [the section 3.1.3]().

### 2.2.2 Installing and uninstalling a package
#### Installing a package with `nest install`
[//]: # (TODO: add link to the section 1.4)
`nest install` is the command to install a package.
To do so, run `nest install <package>` in a terminal.

> Don't forget to use the syntax explained in [the section 1.4]() for the package name!

Reviewer 2: I'm not a huge linguist, but aren't you supposed to say `Use the syntax explained in section 1.4` instead of `in the section 1.4`?

Reviewer: I think the first one is correct. I was actually hesitant to propose this because I figured that the sentence doesn't sound to bad, but if you think so too it's probably a sign it should change.

`nest` will print an error message if more than one package matches the syntax you wrote.
In that case, you should first refine your package name.
To help you with that, you can go on [the official Raven-OS website](https://stable.raven-os.org), which will show you all the packages in the `stable` repository, with their category and all  versions released.

For example, the output of `nest install stable::sys-bin/ncurses6-bin` is as followed:
```
$ nest install ncurses6-bin
4 pending transactions:

 install stable::sys-lib/libncurses6-term#6.1.0
 install stable::sys-lib/libc6#2.28.0
 install stable::sys-lib/libncursestw6#6.1.0
 install stable::sys-bin/ncurses6-bin#6.1.0

Would you like to apply these transactions? [Yes/no]
```

As you can see, the last package listed above is the package you initially asked to install, and above it are all the dependencies that it has and that aren't already installed.

> You will then be prompted with the question `Would you like to apply these transactions?`, which you can answer with either '*yes*', '*Yes*', '*Y*', '*y*', or '*no*', '*No*', '*n*', '*N*'.

Reviewer 2: The enumeration is way too specific and unneeded.

It will remove the given package, its binaries, libraries, manual pages...
Basically everything that was installed with it, but it will keep all the files generated by the package itself.
So, if this package was a text editor for example, the files you created or edited with it won't be altered if you uninstall the package.

However, this may not remove all the package's dependencies, because some of them might still be used by another installed package.

#### Common issues
* `error: no package found for requirement <package>`

This error could mean that you want to install a package that isn't in Nest's cache, so, `nest` can't find it and install it.
Try running `nest pull` and then running your command again.

Another possible explanation is that the package you wanted to install can't be found.
You should check the package name syntax, in case you mistyped the repository, the category, the package name itself or the version.
You can always check [on Raven-OS website](https://stable.raven-os.org/) as mentioned above, to make sure you haven't made any mistakes.

* `error: unable to find an installed package matching <package>`

The package you wanted to uninstall can't be found.

> You have to check the package name syntax for any typo, and that you already have this package installed.

Reviewer 2: ```suggestion
Check the package name syntax for any typo and ensure you actually have this package installed.
```

### 2.3 Summary
This concludes this chapter dedicated to `nest`!

> We saw how to update Nest's cache using `nest pull`, and how to install and to uninstall packages using `nest install` and `nest uninstall` respectively.

Reviewer 2: ```suggestion
We saw how to update Nest's cache using `nest pull`, and how to install and uninstall packages using respectively `nest install` and `nest uninstall`.
```

> Now that you are familiar with `nest`, the next chapter will explain some advanced concepts that are not necessary to be read for the use of `nest`.

Reviewer: "that are not necessary"?
Maybe:
```suggestion
Now that you are familiar with `nest`, the next chapter will explain some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for the use of `nest`.
```
I don't know what the exact link between both chapters is but I think you get the idea.

> Now that you are familiar with `nest`, the next chapter will explain some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for the use of `nest`.

Reviewer 2: ```suggestion
Now that you are familiar with `nest`, the next chapter will explain some advanced concepts related to `nest`, which will help understanding how `nest` works, but aren't strictly needed for a daily use of `nest`.
```

> Readers interested in using `finest` should read the next chapter as well, because those concepts are important in order to use `finest`.

Reviewer 2: ```suggestion
Readers interested in using `finest` should read the next chapter as well.
```

(You have a repetition of the word `use`, and the fact that those concepts are important to use `finest` is implicit in this sentence.

Reviewer: That would make it sound like the next chapter explains finest.

Author: what about
```suggestion
Readers interested in `finest` should read the next chapter as well, because those concepts are important to understand how `finest` works and how to use it.
```

Reviewer: I'd drop the "and how to use it" part.
